### PR TITLE
select

### PR DESCRIPTION
**Pull request by VictorOcio**
add functionality select in inputs to search
is compatible with olders versions.
the new configuration need this type off array:
 $this->Datatable->setTableTypeSearch([
        ['type' => 'input','data' =>''],
        ['type' => 'input','data' =>''],
        ['type'=> 'select','data' =>[
                ['id' => '1','name' => '1'],
                ['id' => '2','name' => '2'],
                ['id' => '3','name' => '3'],
            ]
        ],
        ['type' => 'input','data' =>''],
        ['type' => 'input','data' =>''],
        ['type' => 'input','data' =>''],

### CODE REVIEW

File: src/View/Helper/DatatableHelper.php [hunk 1]
```diff
@@ -48,52 +51,83 @@ class DatatableHelper extends Helper
...
-                    // Search the column for that value
-                    api
-                        .column(colIdx)
-                        .search(
-                            this.value != ''
-                                ? regexr.replace('{search}', '(((' + this.value + ')))')
-                                : '',
-                            this.value != '',
-                            this.value == ''
+        .columns()
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php:62]: tab change here hiding real changes done in the MR, please use the same code style

File: src/View/Helper/DatatableHelper.php [hunk 2]
```diff
@@ -311,16 +350,32 @@ public function getDatatableScript(string $tagId): string
...
 
+        // type of search Headers
+        if ( $this->getConfig('searchHeadersType') !== null) {
+            $this->setTableTypeSearch($this->Config('searchHeadersType'));
+        } elseif ($this->searchHeadersTypes !== null) {
+            $this->setTableTypeSearch($this->searchHeadersTypes);
+        } else {
+            throw new MissConfiguredException(__('Search headers type not configured'));
+        }
+
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php]: remove estra lines

File: src/View/Helper/DatatableHelper.php [hunk 3]
```diff
@@ -311,16 +350,32 @@ public function getDatatableScript(string $tagId): string
...
+     
+  
         $this->processColumnRenderCallbacks();
         $this->processColumnDefinitionsCallbacks();
+        $this->searchHeadersTypes= $this->processColumnTypeSearch();
+       
         $this->validateConfigurationOptions();
 
+
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php]: remove extra blank lines

File: src/View/Helper/DatatableHelper.php [hunk 4]
```diff
@@ -311,16 +350,32 @@ public function getDatatableScript(string $tagId): string
...
+        $this->columnSearchTemplate = sprintf($this->columnSearchTemplate, $this->searchHeadersTypes);
+
         if ($this->getConfig('columnSearch')) {
             $columnSearchTemplate = sprintf($this->columnSearchHeaderTemplate, $tagId, $tagId);
         } else {
             $columnSearchTemplate = '';
         }
-
+
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php]: remove

File: src/View/Helper/DatatableHelper.php [hunk 5]
```diff
@@ -363,6 +418,38 @@ protected function validateConfigurationOptions()
...
+    /**
+     * Loop types into javascript format.
+     */
+    protected function processColumnTypeSearch() {
+
+        $rows = [];
+        foreach ($this->searchHeadersTypes as $definition) {
+            $parts = [];
+            
+                foreach ($definition as $key => $val) {
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php]: don't use $key $val in foreach, assign proper names to help readability

File: src/View/Helper/DatatableHelper.php [hunk 6]
```diff
@@ -374,12 +461,13 @@ protected function processExtraFields()
             foreach ($definition as $key => $val) {
                 $parts[] = "'{$key}': {$val}";
             }
-            $rows[] = implode(',', $parts);
+            $rows[] = implode(',', $parts);
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php]: remove spaces

File: src/View/Helper/DatatableHelper.php [hunk 7]
```diff
@@ -374,12 +461,13 @@ protected function processExtraFields()
...
                 $parts[] = "'{$key}': {$val}";
             }
-            $rows[] = implode(',', $parts);
+            $rows[] = implode(',', $parts);   
         }
-
         return implode(',', $rows);
     }
 
+
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php]: remove

File: src/View/Helper/DatatableHelper.php [hunk 8]
```diff
@@ -502,4 +590,40 @@ public function getTableHeaders(
...
         return $this->Html->tableHeaders($tableHeaders, $headersAttrsTr, $headersAttrsTh);
     }
+
+    /**
+     * Put Definition of types of search in headers
+     * 
+     * @param iterable|null $tableSearchHeaders
+     * 
+     */
+
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php]: remove

Comment (steinkel) [reply src/View/Helper/DatatableHelper.php]: please install and use cakephp codesniffer, cleanup all the files

File: src/View/Helper/DatatableHelper.php [hunk 9]
```diff
@@ -12,6 +12,8 @@
 use Datatables\Exception\MissConfiguredException;
 use InvalidArgumentException;
 
+use function PHPUnit\Framework\throwException;
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php]: why is this needed?

File: src/View/Helper/DatatableHelper.php [hunk 10]
```diff
@@ -12,6 +12,8 @@
 use Datatables\Exception\MissConfiguredException;
```
Comment (steinkel) [inline src/View/Helper/DatatableHelper.php]: write documentation about how to use the feature, add examples and limitations to the feature

**Review by steinkel — CHANGES_REQUESTED** [review]

**Review by VictorOcio — COMMENTED** [review]
hi, i think to resolve the problems.
i can put a image in readme.md?
